rhizome/host/lib/vm_setup.rb

```diff
@@ -44,6 +44,7 @@ def vp
   end
 
   def prep(unix_user, public_key, nics, gua, ip4, local_ip4, boot_image, max_vcpus, cpu_topology, mem_gib, ndp_needed, storage_volumes, storage_secrets, swap_size_bytes)
+    stop_services
```

Comment:
when does it happen the services we know about are both started and prep gets run? 

From what I can tell, this is the only site where VMs get started, and I believe the dnsmasq component gets started via a dependency:

`prog/vm/nexus.rb`
```ruby
  label def run
    host.sshable.cmd("sudo systemctl start #{q_vm}")
    hop_wait_sshable
  end
```

Notably, it only happens after prep succeeds on the prog side:


```ruby
  label def create_unix_user
    host.sshable.cmd("sudo userdel --remove --force #{q_vm} || true")
    host.sshable.cmd("sudo groupdel -f #{q_vm} || true")

    # create vm's user and home directory
    uid = rand(1100..59999)
    host.sshable.cmd("sudo adduser --disabled-password --gecos '' --home #{vm_home.shellescape} --uid #{uid} #{q_vm}")

    hop_prep
  end

  label def prep
    case host.sshable.cmd("common/bin/daemonizer --check prep_#{q_vm}")
    when "Succeeded"
      host.sshable.cmd("common/bin/daemonizer --clean prep_#{q_vm}")
      vm.nics.each { _1.incr_setup_nic }
      bud Prog::Vnet::UpdateFirewallRules, {subject_id: vm.id}, :update_firewall_rules
      hop_wait_firewall_rules_before_run
    when "NotStarted", "Failed"
      topo = vm.cloud_hypervisor_cpu_topology

      # we don't write secrets to params_json, because it
      # shouldn't be stored in the host for security reasons.
      params_json = JSON.pretty_generate({
        "vm_name" => vm_name,
        "public_ipv6" => vm.ephemeral_net6.to_s,
        "public_ipv4" => vm.ip4.to_s || "",
        "local_ipv4" => local_ipv4,
        "unix_user" => vm.unix_user,
        "ssh_public_key" => vm.public_key,
        "nics" => vm.nics.map { |nic| [nic.private_ipv6.to_s, nic.private_ipv4.to_s, nic.ubid_to_tap_name, nic.mac] },
        "boot_image" => vm.boot_image,
        "max_vcpus" => topo.max_vcpus,
        "cpu_topology" => topo.to_s,
        "mem_gib" => vm.mem_gib,
        "ndp_needed" => host.ndp_needed,
        "storage_volumes" => storage_volumes,
        "swap_size_bytes" => frame["swap_size_bytes"]
      })

      secrets_json = JSON.generate({
        storage: storage_secrets
      })

      # Enable KVM access for VM user.
      host.sshable.cmd("sudo usermod -a -G kvm #{q_vm}")

      # put prep.json
      host.sshable.cmd("sudo -u #{q_vm} tee #{params_path.shellescape}", stdin: params_json)

      host.sshable.cmd("common/bin/daemonizer 'sudo host/bin/prepvm.rb #{params_path.shellescape}' prep_#{q_vm}", stdin: secrets_json)
    end

    nap 5
  end

  label def wait_firewall_rules_before_run
    reap
    hop_run if leaf?
    donate
  end
```

And we know that VMs stuck at `create_unix_user` is implicated. Notably this is the first label after allocation in `start` where the prog starts to try to manipulate the VmHost at all to create the Vm.

Do you have a viable theory how process leakage issues are related here? The Vm hasn't really gotten an opportunity to "do" anything. No systemctl start should been isued yet.

Here's an alternate theory:

recreate-persisted, and it alone, has a bug. The bug could be of this format:

1. create the ip link
2. crash on some later step
3. when we try to run again, it crashes on create ip link

We mostly know about `#3`, but my guess is the bug is in `#2` and `#3` shows up in repeats. That's the evidence in the contemporary notes. `#2` may have something to do with the VM having an especially brief life.

Going back to the Mezmo logs, we don't keep stdout/stderr from ssherrors that blow the stack, fortunately/unfortunately (it would have been nice to have but also could be loaded with a secret in the general case).

One semi-solution to this is we should probably stop doing this:

```
    command = "/bin/bash -c '#{command} > #{dir}/var/log/#{name}.stdout 2> #{dir}/var/log/#{name}.stderr'"
```

And instead put things in the journal, so that we can get the first copy of a fault...at least for a limited time. This version guarantees we only see the last fault (which may be independently useful, but if I had to choose one, I'd choose the journal). In this case I probably wouldn't have had the wherewithal to look, though. But the problem of "bug exactly once, retry different" seems like a viable fault here.

Reply:
Okay, after going through this, I now understand https://github.com/ubicloud/ubicloud/pull/1394 more completely. We should solve it closer to that way.

There are two ways to go about that I think:

1. Don't run recreate-unpersisted when unnecessary at all. the pre-condition state must be that the VM is not running. 

2. recreate-unpersisted can check that the relevant objects are created (at least some of them, that can be an indicator) and indicate success.

I think on the whole I prefer #1. We know this is a fresh VM that needs no work to be done, so let's not do any work. But the robustness to code change and/or evaluation under concurrency needs some work.